File: stackunderflow/etl/normalize/cline.py
"""Cline (and Cline-family) normalizer.

Cline writes one ``ui_messages.json`` per task containing a stream of
events; the adapter (``stackunderflow/adapters/cline.py``) splits that
stream into one ``Record`` per ``api_req_started`` event. Each Record
lands as a single row in the ``messages`` table — i.e. the per-event
grain is preserved at the storage layer, not aggregated to the task.

The spec ("Cline persists per-task, not per-message") refers to the
on-disk source-of-truth, not the table grain we ingest into. Per
``docs/specs/etl-architecture.md`` Wave 2A:

    The adapter already emits per-``api_req_started`` events; preserve
    that grain. Tokens from the ``api_req_started.text`` JSON.

So one msg_row → one event. A task with 3 ``api_req_started`` events
produces 3 messages-table rows, which normalize into 3 usage_events
rows — one per call.

Token shape — ``api_req_started.text`` is a JSON-stringified blob
``{tokensIn, tokensOut, cacheWrites, cacheReads, cost}``. We parse
that here so the normalizer is the authoritative source even when an
upgrade path leaves stale column values on the row. The pricer-side
``cost`` field (which Cline pre-computes against its own rate table)
is preserved in ``raw_extras`` for cross-reference but **not** used —
we recompute via ``compute_cost`` for parity with the rest of the
pipeline.

Cline runs against the Anthropic API directly (the user's own key), so
the canonical 4-token shape applies as-is — ``cacheWrites`` →
``cache_create``, ``cacheReads`` → ``cache_read``.

Model resolution — Cline declares the model once on the first user
message via ``<model>...</model>``. The adapter lifts that onto every
``Record.model``, so the messages-table row already carries it. We
trust the column.
"""

from __future__ import annotations

import json
from collections.abc import Iterable

from stackunderflow.infra.costs import RATE_CARD

from .base import (
    COST_SOURCE_RATE_CARD,
    COST_SOURCE_UNKNOWN,
    Normalizer,
)

# Default model — keep aligned with the adapter's _DEFAULT_MODEL.
_DEFAULT_MODEL = "cline-auto"

# Cline embeds the upstream-computed cost on every api_req_started
# event; we preserve it for debugging but do not consume it.
_RAW_EXTRAS_FIELDS = ("cost", "request", "apiProtocol")


class ClineNormalizer(Normalizer):
    provider_name = "cline"

    def normalize(self, msg_row: dict) -> Iterable[dict]:
        role = str(msg_row.get("role") or "")
        if role != "assistant":
            return

        tokens = _parse_api_req_tokens(msg_row)
        if tokens is None:
            return

        if (
            tokens["input"] == 0
            and tokens["output"] == 0
            and tokens["cache_read"] == 0
            and tokens["cache_create"] == 0
        ):
            return

        model = str(msg_row.get("model") or "") or _DEFAULT_MODEL

        cost_source = (
            COST_SOURCE_RATE_CARD
            if model in RATE_CARD
            else COST_SOURCE_UNKNOWN
        )

        raw_extras = _extras_from_raw_json(msg_row.get("raw_json"))

        yield self._build_event(
            msg_row,
            input_tokens=tokens["input"],
            output_tokens=tokens["output"],
            cache_read_tokens=tokens["cache_read"],
            cache_create_tokens=tokens["cache_create"],
            cost_source=cost_source,
            model=model,
            raw_extras=raw_extras,
        )


# ── helpers ─────────────────────────────────────────────────────────


def _parse_api_req_tokens(msg_row: dict) -> dict[str, int] | None:
    """Return canonical 4-key tokens parsed from the api_req_started event.

    Resolution order:
      1. The event's ``text`` field (a JSON-stringified blob) reachable
         via ``raw_json.text``. This is the on-disk source of truth.
      2. Direct ``text`` field on msg_row (synthetic test fixtures
         can pass it without an enclosing raw_json).
      3. Pre-canonicalised messages-table columns the adapter wrote.

    Returns None when the row carries no usage data at all (e.g. an
    api_req_started event whose text payload was malformed). User
    feedback / text events are filtered upstream by the role check
    and never reach this helper.
    """
    text_field = _extract_text_field(msg_row)
    if text_field is not None:
        parsed = _safe_parse_json(text_field)
        if isinstance(parsed, dict):
            return _canonicalize(parsed)

    # Fall back to the columns the adapter already wrote.
    return {
        "input": int(msg_row.get("input_tokens") or 0),
        "output": int(msg_row.get("output_tokens") or 0),
        "cache_read": int(msg_row.get("cache_read_tokens") or 0),
        "cache_create": int(msg_row.get("cache_create_tokens") or 0),
    }


def _extract_text_field(msg_row: dict) -> str | None:
    """Return the ``api_req_started.text`` JSON string, or None."""
    direct = msg_row.get("text")
    if isinstance(direct, str) and direct:
        return direct
    payload = _safe_load_raw(msg_row.get("raw_json"))
    if isinstance(payload, dict):
        text = payload.get("text")
        if isinstance(text, str) and text:
            return text
    return None


def _canonicalize(parsed: dict) -> dict[str, int]:
    """Map Cline's tokens dict to the canonical 4-key shape."""
    return {
        "input": _safe_int(parsed.get("tokensIn")),
        "output": _safe_int(parsed.get("tokensOut")),
        "cache_read": _safe_int(parsed.get("cacheReads")),
        "cache_create": _safe_int(parsed.get("cacheWrites")),
    }


def _safe_int(value: object) -> int:
    try:
        return max(int(value or 0), 0)
    except (TypeError, ValueError):
        return 0


def _extras_from_raw_json(raw_json: object) -> dict | None:
    payload = _safe_load_raw(raw_json)
    if not isinstance(payload, dict):
        return None

    # Cline's payload may itself be the ``api_req_started`` event, whose
    # ``text`` carries the cost. Try to surface ``cost`` either way.
    out: dict = {}
    parsed_text = _safe_parse_json(payload.get("text"))
    if isinstance(parsed_text, dict):
        cost = parsed_text.get("cost")
        if cost is not None:
            out["cost"] = cost
        for key in _RAW_EXTRAS_FIELDS:
            if key in parsed_text and key != "cost":
                out[key] = parsed_text[key]

    for key in _RAW_EXTRAS_FIELDS:
        val = payload.get(key)
        if val is not None and key not in out:
            out[key] = val

    return out or None


def _safe_load_raw(raw_json: object) -> object | None:
    if isinstance(raw_json, dict):
        return raw_json
    if not isinstance(raw_json, str | bytes | bytearray):
        return None
    try:
        if isinstance(raw_json, bytes | bytearray):
            return json.loads(raw_json.decode("utf-8", errors="replace"))
        return json.loads(raw_json)
    except (json.JSONDecodeError, ValueError, UnicodeDecodeError):
        return None


def _safe_parse_json(text: object) -> object | None:
    if not isinstance(text, str) or not text:
        return None
    try:
        return json.loads(text)
    except (json.JSONDecodeError, ValueError):
        return None

File: rust/crates/stax-etl/src/normalize/cline.rs
//! Cline and its two forks — port of `etl/normalize/cline.py`,
//! `kilocode.py` and `roocode.py`.
//!
//! **One transform, three registry keys.** `KiloCodeNormalizer` and
//! `RooCodeNormalizer` are `class X(ClineNormalizer)` bodies that override
//! nothing but `provider_name`, so this module is one struct carrying the key
//! as a field rather than three copies of the parser. The key is not cosmetic:
//! it is the pricer key `_compute_cost_usd` routes through
//! (`provider=self.provider_name`), and it is the `provider` column's fallback.
//!
//! The token source of truth is the `api_req_started.text` blob — a
//! JSON-*stringified* `{tokensIn, tokensOut, cacheWrites, cacheReads, cost}`
//! nested inside `raw_json` — parsed here so the normalizer stays authoritative
//! even when an upgrade path left stale column values on the row. Cline's own
//! pre-computed `cost` is preserved in `raw_extras` for cross-reference and
//! deliberately **not** consumed: every provider re-prices through
//! `compute_cost` so the marts read one number computed one way.

use stax_core::queries::pyjson::Value as PyValue;

use super::base::{EventSpec, NormalizeContext, Normalizer, UsageEvent};
use super::row::{
    MsgRow, PyRaise, as_nonempty_str, dict_get, int_or_zero, safe_int, safe_load_raw,
    safe_parse_json, str_or_empty,
};

/// Keep aligned with the adapter's `_DEFAULT_MODEL`.
const DEFAULT_MODEL: &str = "cline-auto";

/// Cline embeds the upstream-computed cost on every `api_req_started` event;
/// preserved for debugging, never consumed.
const RAW_EXTRAS_FIELDS: [&str; 3] = ["cost", "request", "apiProtocol"];

/// The canonical four tokens as `_parse_api_req_tokens` returns them.
struct ClineTokens {
    input: i64,
    output: i64,
    cache_read: i64,
    cache_create: i64,
}

/// The Cline-family transform. One struct, three registered keys.
#[derive(Debug, Clone, Copy)]
pub struct ClineNormalizer {
    provider: &'static str,
}

impl ClineNormalizer {
    /// `provider_name = "cline"`.
    #[must_use]
    pub const fn cline() -> Self {
        Self { provider: "cline" }
    }

    /// `class KiloCodeNormalizer(ClineNormalizer)` — same transform, own key,
    /// own pricer route (the pricer map sends `kilocode` to Anthropic).
    #[must_use]
    pub const fn kilocode() -> Self {
        Self {
            provider: "kilocode",
        }
    }

    /// `class RooCodeNormalizer(ClineNormalizer)`.
    #[must_use]
    pub const fn roocode() -> Self {
        Self {
            provider: "roocode",
        }
    }
}

impl Normalizer for ClineNormalizer {
    fn provider_name(&self) -> &'static str {
        self.provider
    }

    fn normalize(&self, ctx: &NormalizeContext, row: &MsgRow) -> Result<Vec<UsageEvent>, PyRaise> {
        if str_or_empty(row, "role") != "assistant" {
            return Ok(Vec::new());
        }

        // `_parse_api_req_tokens` returns `None` only in the docstring: every
        // branch returns a dict, so the `if tokens is None: return` guard below
        // it is dead on arrival. Ported as written — the shape is what a future
        // edit would restore, and pretending it does not exist would hide that.
        let tokens = parse_api_req_tokens(row)?;

        if tokens.input == 0
            && tokens.output == 0
            && tokens.cache_read == 0
            && tokens.cache_create == 0
        {
            return Ok(Vec::new());
        }

        let model = match str_or_empty(row, "model") {
            empty if empty.is_empty() => DEFAULT_MODEL.to_string(),
            model => model,
        };
        let cost_source = ctx.rate_card_or_unknown(&model);
        let raw_extras = extras_from_raw_json(row.get("raw_json"));

        Ok(vec![
            self.build_event(
                ctx,
                row,
                EventSpec::new(
                    tokens.input,
                    tokens.output,
                    tokens.cache_read,
                    tokens.cache_create,
                    cost_source,
                )
                .model(model)
                .raw_extras(raw_extras),
            ),
        ])
    }
}

/// Canonical four tokens from the `api_req_started` event.
///
/// Resolution order: the event's stringified `text` payload (on-disk truth),
/// then a `text` field passed directly by a synthetic row, then the columns the
/// adapter already wrote.
fn parse_api_req_tokens(row: &MsgRow) -> Result<ClineTokens, PyRaise> {
    let parsed = extract_text_field(row).and_then(|text| stax_core::queries::pyjson::loads(&text));
    if let Some(parsed) = parsed.as_ref()
        && matches!(parsed, PyValue::Object(_))
    {
        return Ok(ClineTokens {
            input: safe_int(parsed.get("tokensIn"))?,
            output: safe_int(parsed.get("tokensOut"))?,
            cache_read: safe_int(parsed.get("cacheReads"))?,
            cache_create: safe_int(parsed.get("cacheWrites"))?,
        });
    }
    Ok(ClineTokens {
        input: int_or_zero(row, "input_tokens")?,
        output: int_or_zero(row, "output_tokens")?,
        cache_read: int_or_zero(row, "cache_read_tokens")?,
        cache_create: int_or_zero(row, "cache_create_tokens")?,
    })
}

/// The `api_req_started.text` JSON string, or `None`.
fn extract_text_field(row: &MsgRow) -> Option<String> {
    if let Some(direct) = as_nonempty_str(row.get("text")) {
        return Some(direct.to_string());
    }
    let payload = safe_load_raw(row.get("raw_json"))?;
    let text = dict_get(&payload, "text")?;
    as_nonempty_str(Some(text)).map(ToString::to_string)
}

/// `cost` (from the nested `text` blob or the payload) plus `request` /
/// `apiProtocol`, in the order Python inserts them.
fn extras_from_raw_json(raw_json: Option<&PyValue>) -> Option<PyValue> {
    let payload = safe_load_raw(raw_json)?;
    if !matches!(payload, PyValue::Object(_)) {
        return None;
    }

    let mut out: Vec<(String, PyValue)> = Vec::new();
    let parsed_text = safe_parse_json(payload.get("text"));
    if let Some(parsed_text) = parsed_text.as_ref()
        && matches!(parsed_text, PyValue::Object(_))
    {
        if let Some(cost) = parsed_text.get("cost")
            && *cost != PyValue::Null
        {
            out.push(("cost".to_string(), cost.clone()));
        }
        for key in RAW_EXTRAS_FIELDS {
            if key == "cost" {
                continue;
            }
            // `if key in parsed_text` — presence, so an explicit `null` counts.
            if let Some(value) = parsed_text.get(key) {
                out.push((key.to_string(), value.clone()));
            }
        }
    }

    for key in RAW_EXTRAS_FIELDS {
        let Some(value) = payload.get(key) else {
            continue;
        };
        if *value == PyValue::Null || out.iter().any(|(name, _)| name == key) {
            continue;
        }
        out.push((key.to_string(), value.clone()));
    }

    (!out.is_empty()).then_some(PyValue::Object(out))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::normalize::base::CostSource;
    use crate::normalize::test_support::{assistant_row, ctx};

    fn cline_row(text: &str) -> MsgRow {
        assistant_row("cline", "claude-sonnet-4-5-20250929").with(
            "raw_json",
            PyValue::Str(format!(r#"{{"text": {}}}"#, json_string(text))),
        )
    }

    fn json_string(text: &str) -> String {
        stax_core::queries::pyjson::dumps_compact(&PyValue::Str(text.to_string()))
    }

    #[test]
    fn tokens_come_from_the_stringified_text_blob_not_the_columns() {
        let row = cline_row(
            r#"{"tokensIn": 111, "tokensOut": 22, "cacheReads": 3, "cacheWrites": 4, "cost": 0.5}"#,
        )
        // Stale columns the parser must ignore.
        .with("input_tokens", PyValue::Int(999_999));
        let events = ClineNormalizer::cline().normalize(&ctx(), &row).unwrap();
        assert_eq!(events.len(), 1);
        assert_eq!(events[0].input_tokens, 111);
        assert_eq!(events[0].output_tokens, 22);
        assert_eq!(events[0].cache_read_tokens, 3);
        assert_eq!(events[0].cache_create_tokens, 4);
        // The upstream cost is preserved, not consumed.
        assert_eq!(events[0].raw_extras.as_deref(), Some(r#"{"cost": 0.5}"#));
        assert!(events[0].cost_usd > 0.0);
    }

    #[test]
    fn a_malformed_text_blob_falls_back_to_the_columns() {
        let row = cline_row("not json at all")
            .with("input_tokens", PyValue::Int(10))
            .with("output_tokens", PyValue::Int(20));
        let events = ClineNormalizer::cline().normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].input_tokens, 10);
        assert_eq!(events[0].output_tokens, 20);
    }

    #[test]
    fn negative_wire_counts_clamp_to_zero_and_can_zero_the_whole_row() {
        let row = cline_row(r#"{"tokensIn": -5, "tokensOut": -5}"#);
        assert!(
            ClineNormalizer::cline()
                .normalize(&ctx(), &row)
                .unwrap()
                .is_empty()
        );
    }

    #[test]
    fn the_default_model_stands_in_for_a_blank_column() {
        let row = cline_row(r#"{"tokensIn": 5}"#).with("model", PyValue::Null);
        let events = ClineNormalizer::cline().normalize(&ctx(), &row).unwrap();
        assert_eq!(events[0].model, DEFAULT_MODEL);
        // `cline-auto` is a first-party rate-card id, so this is not `unknown`.
        assert_eq!(events[0].cost_source, CostSource::RateCard);
    }

    #[test]
    fn the_three_keys_share_one_transform_and_differ_only_in_routing() {
        let row = cline_row(r#"{"tokensIn": 100, "tokensOut": 10}"#);
        let ctx = ctx();
        let cline = ClineNormalizer::cline().normalize(&ctx, &row).unwrap();
        let kilo = ClineNormalizer::kilocode().normalize(&ctx, &row).unwrap();
        let roo = ClineNormalizer::roocode().normalize(&ctx, &row).unwrap();
        for other in [&kilo, &roo] {
            assert_eq!(other[0].input_tokens, cline[0].input_tokens);
            assert_eq!(other[0].output_tokens, cline[0].output_tokens);
            // Cline-family extensions all run against the user's Anthropic key,
            // so the pricer map routes all three to the same rates.
            assert_eq!(other[0].cost_usd.to_bits(), cline[0].cost_usd.to_bits());
        }
        // The provider column comes off the ROW, so it agrees here; the
        // difference the keys make is the pricer route and the fallback.
        assert_eq!(
            ClineNormalizer::kilocode()
                .normalize(&ctx, &row.clone().with("provider", PyValue::Null))
                .unwrap()[0]
                .provider,
            "kilocode"
        );
    }

    #[test]
    fn non_assistant_rows_yield_nothing() {
        let row = cline_row(r#"{"tokensIn": 100}"#).with("role", PyValue::Str("user".into()));
        assert!(
            ClineNormalizer::cline()
                .normalize(&ctx(), &row)
                .unwrap()
                .is_empty()
        );
    }
}
